Create mqtt eventstream component

moonshot opened the pull request:
@balloob are you accepting PRs for new components?

My home is aboard a boat in a marina. My network connects to the marina's network via a wireless AP in client mode, and I have no control over port forwarding of the marina's router. So the HASS cannot accept incoming HTTP traffic.

I built on top of the MQTT component so that the boat's HA instance can communicate via a cloud MQTT broker to a HA instance that I have running on Heroku.

I wasn't really sure what to name this or where to put it. Let me know what you recommend.

moonshot committed:
Create mqtt eventstream component

balloob commented on homeassistant/components/mqtt_eventstream.py:
```diff
@@ -0,0 +1,93 @@
+            return
+        if event.event_type in (
+                EVENT_TIME_CHANGED,
+                EVENT_CALL_SERVICE,
```
Why filter out service calls? Wouldn't it be cool to be able to call remote services?

moonshot replied:
Ah, right, I had meant to flag this for review or add a TODO.

Need to figure out the code for filtering out the service call made by the mqtt component publishing the event, otherwise you get into an infinite loop. I had filtered out all EVENT_CALL_SERVICE and EVENT_SERVICE_EXECUTED events to get around that for now.

I'll tag you after I do both that and verify the code manually with 2 instances.

balloob replied:
I am confused. There is no service call for publishing the event right?

moonshot replied:
Oh, sorry for being unclear. I meant the [mqtt domain's publish event](https://github.com/balloob/home-assistant/blob/dev/homeassistant/components/mqtt/__init__.py#L133) which gets triggered by publishing the event to the broker for the remote HA to pick up. If that makes any sense...

balloob replied:
Ah yes, that is a tricky one. As this component is about event forwarding and very closely coupled to the MQTT component, it _might_ make sense to not use the publish service but use the internal method directly:

```python
from homeassistant.components import mqtt
mqtt.MQTT_CLIENT.publish(msg_topic, payload, qos, retain)
```

This will also prevent the history component from recording every event twice.

I am actually not sure if I like this because it goes against the core principle that the event bus is the history of everything happening in the system.

moonshot replied:
@balloob I think I was able to get a reasonable solution this way:
* filter out call service events for publishing to the configured topic
* filter out all service executed calls

balloob commented:
This is a very nice addition, I like it! One minor comment but the code looks very good.

I think the name and it being a component is fine.

moonshot committed:
Fix filtering of EVENT_CALL_SERVICE and EVENT_SERVICE_EXECUTED events

moonshot commented:
@balloob 
I added another commit for review so this PR should be good to go code-wise.
Looks like the travis build is failing because of pylint violations in other files. I just rebased dev but no dice. :(

balloob commented:
Don't worry about those, we got hit by a linter that issued an update and got more strict. Happens from time to time but we'll end up eventually in a better place so it's ok.
